libtenzir/src/curl.cpp

```diff
@@ -0,0 +1,173 @@
+  curl_multi_remove_handle(multi_, easy_);
+  curl_easy_cleanup(easy_);
+  curl_multi_cleanup(multi_);
+}
+
+auto curl::set(const curl_options& opts) -> caf::error {
+  if (not opts.default_protocol.empty())
+    if (auto err = set(CURLOPT_DEFAULT_PROTOCOL, opts.default_protocol.c_str()))
+      return err;
+  if (not opts.url.empty())
+    if (auto err = set(CURLOPT_URL, opts.url.c_str()))
+      return err;
+  if (opts.default_protocol == "http" || opts.default_protocol == "https") {
+    if (not opts.http.body.empty() || opts.http.method == "POST") {
+      // Setting the size manually is important because curl would otherwise
+      // use strlen to determine the body size. We may pass binary data at
+      // some point.
+      auto size = detail::narrow_cast<long>(opts.http.body.size());
+      if (auto err = set(CURLOPT_POSTFIELDSIZE, size))
+        return err;
+      // Here is choice of CURLOPT_POSTFIELDS vs. CURLOPT_COPYPOSTFIELDS,
+      // where the latter copies the input. We're copying now, but should
+      // revisit this for large POST bodies.
+      if (auto err = set(CURLOPT_COPYPOSTFIELDS, opts.http.body.c_str()))
+        return err;
+      // NB: setting body data via CURLOPT_POSTFIELDS implicitly sets the
+      // Content-Type header to 'application/x-www-form-urlencoded' unless we
+      // provide a Content-Type header that overrides it.
+    }
+    if (opts.http.method == "GET") {
+      if (auto err = set(CURLOPT_HTTPGET, 1L))
+        return err;
+    } else if (opts.http.method == "POST") {
+      if (auto err = set(CURLOPT_POST, 1L))
+        return err;
+    } else if (opts.http.method == "HEAD") {
+      if (auto err = set(CURLOPT_NOBODY, 1L))
+        return err;
+    } else if (not opts.http.method.empty()) {
+      if (auto err = set(CURLOPT_CUSTOMREQUEST, opts.http.method.c_str()))
+        return err;
+    }
+    if (not opts.http.headers.empty()) {
+      if (headers_) {
+        curl_slist_free_all(headers_);
+        headers_ = nullptr;
+      }
+      for (const auto& [header, value] : opts.http.headers) {
+        auto str = header;
+        str += ':';
+        if (not value.empty()) {
+          str += ' ';
+          str += value;
+        }
+        headers_ = curl_slist_append(headers_, str.c_str());
+      }
+    }
+    if (auto err = set(CURLOPT_HTTPHEADER, headers_))
+      return err;
+  }
+  return {};
+}
+
+auto curl::download(std::chrono::milliseconds timeout) -> generator<chunk_ptr> {
+  // curl_easy_setopt(easy_, CURLOPT_VERBOSE, 1L);
+  std::vector<chunk_ptr> chunks;
+  auto err = set(CURLOPT_WRITEDATA, &chunks);
+  TENZIR_ASSERT(not err);
+  auto ms = detail::narrow_cast<int>(timeout.count());
+  auto still_running = int{0};
+  do {
+    auto mc = curl_multi_perform(multi_, &still_running);
+    if (mc == CURLM_OK && still_running != 0)
+      mc = curl_multi_poll(multi_, nullptr, 0u, ms, nullptr);
+    if (mc != CURLM_OK) {
+      TENZIR_ERROR(to_error(mc));
+      co_return;
```

Comment:
When I enter an unresolvable URL with correct syntax, `curl` will report:
```
Could not resolve host
```
but `mc` will still be `CURLM_OK`.

I think just relying on `CURLMCode` for error reporting is a bit too naive to be sustainable (see also the URL parsing comment) .

Reply:
We will leave this as-is for now and wait for user feedback.

Reply:
I lifted the error handling in e83f661aec, but even though DNS resolution fails, `curl_multi_perform` still returns "okay". I'm not sure how to proceed, as a quick Google search didn't yield any insight.

Let's revisit this when it becomes a bigger problem.

Reply:
As someone who didn't know about this behavior and just stumbled upon it, I was very confused. I created https://github.com/tenzir/issues/issues/986 for this.